### PR TITLE
SpecTf model support

### PR DESCRIPTION
**Pull request by jakehlee**
_In progress:_
- [x] Remove `*out` wildcard from gitignore
- [x] Remove research configurations
- [x] Test `report_spectf.py`
- [x] README for SpecTf setup

Changelog:

`training_spectf.py`
* This is the SpecTf training script that relies on https://github.com/emit-sds/spectf

`report_spectf.py`
* This generates a report using a pretrained model, and is also meant to serve as an inference example.

`training_example.py`
* Added GPU support
* Fixed f-string quotation

`metrics.py`
* Added function to determine optimal f1 threshold. 
* `generate_metrics()` now uses the optimal f1 threshold if no such threshold is specified.

### CODE REVIEW

**Comment by jakehlee** [comment]
(Ignoring comments about test coverage since reporting/metrics don't have test coverage anyways)

File: models/spectf/training_spectf.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    help="Path to the YAML config for the dataloader.",
+    envvar=f'{ENV_VAR_PREFIX}_MODEL_CONFIG'
+)
+@click.option(
+    "--simulated-test-set-size",
+    required=False,
+    type=int,
+    default=100_000,
+    help="Number of rows to generate for the simulated test set.",
+    envvar=f'{ENV_VAR_PREFIX}_SIMULATED_TEST_SET_SIZE'
```
[the remark this answers is not in the document]
Comment (jakehlee) [reply models/spectf/training_spectf.py:70]: Following convention in training_example.py

File: models/spectf/report_spectf.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    type=click.Path(exists=True, dir_okay=True, file_okay=False),
+    help="Output file directory.",
+    envvar=f'{ENV_VAR_PREFIX}OUTDIR'
+)
+@click.option(
+    "--data-config",
+    required=True,
+    type=click.Path(exists=True, dir_okay=False),
+    help="Path to the YAML config for the dataloader.",
+    envvar=f'{ENV_VAR_PREFIX}_DATA_CONFIG'
```
[the remark this answers is not in the document]
Comment (jakehlee) [reply models/spectf/report_spectf.py:53]: training_example.py

File: src/cover_class/reporting/metrics.py [hunk 3]
```diff
@@ -192,3 +192,43 @@ def brier_score(): ...
...
+            fp = int(((yt == 0) & (yp_bin == 1)).sum())
+            fn = int(((yt == 1) & (yp_bin == 0)).sum())
+            prec = tp / (tp + fp + 1e-12)
+            rec  = tp / (tp + fn + 1e-12)
+            fb   = (1 + beta**2) * prec * rec / (beta**2 * prec + rec + 1e-12)
+            if fb > best_fb:
+                best_fb  = fb
+                best_thr = float(thr)
+        opt_thr.append(best_thr)
+    return opt_thr
```
[the remark this answers is not in the document]
Comment (jakehlee) [reply src/cover_class/reporting/metrics.py:234]: ignoring test coverage

File: models/spectf/training_spectf.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    type=click.Path(exists=True, dir_okay=True, file_okay=False),
+    help="Output file directory.",
+    envvar=f'{ENV_VAR_PREFIX}OUTDIR'
+)
+@click.option(
+    "--data-config",
+    required=True,
+    type=click.Path(exists=True, dir_okay=False),
+    help="Path to the YAML config for the dataloader.",
+    envvar=f'{ENV_VAR_PREFIX}_DATA_CONFIG'
```
[the remark this answers is not in the document]
Comment (jakehlee) [reply models/spectf/training_spectf.py:55]: training_example.py

File: models/spectf/training_spectf.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+)
+@click.option(
+    "--model-config",
+    required=True,
+    type=click.Path(exists=True, dir_okay=False),
+    help="Path to the YAML config for the dataloader.",
+    envvar=f'{ENV_VAR_PREFIX}_MODEL_CONFIG'
```
[the remark this answers is not in the document]
Comment (jakehlee) [reply models/spectf/training_spectf.py:62]: training_example.py

File: models/spectf/report_spectf.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    type=click.Path(exists=True, dir_okay=False),
+    help="Path to the YAML config for the model.",
+    envvar=f'{ENV_VAR_PREFIX}_MODEL_CONFIG'
+)
+@click.option(
+    "--model-weights",
+    required=True,
+    type=click.Path(exists=True, dir_okay=False),
+    help="Path to the model weights file (.pth).",
+    envvar=f'{ENV_VAR_PREFIX}_MODEL_WEIGHTS'
```
[the remark this answers is not in the document]
Comment (jakehlee) [reply models/spectf/report_spectf.py:67]: training_example.py

File: models/spectf/training_spectf.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        # Calculate test set metrics using the best thresholds for the test set
+        _figs = []
+        test_metrics = report.generate_metrics(simulation_y_test, y_hat, None, _figs, class_names)
+        for f in _figs: plt.close(f)
+        del _figs
+
+        # Extract the thresholds used for the test set 
+        test_thresholds = [test_metrics[class_name]['Threshold'] for class_name in class_names]
+
+        # Calculate OOD validation set metrics using the test set thresholds
```
[the remark this answers is not in the document]
Comment (jakehlee) [reply models/spectf/training_spectf.py:247]: test set / validation set definitions are flipped in our case

File: models/spectf/report_spectf.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+)
+@click.option(
+    "--simulated-test-set-size",
+    required=False,
+    type=int,
+    default=100_000,
+    help="Number of rows to generate for the simulated test set.",
+    envvar=f'{ENV_VAR_PREFIX}_SIMULATED_TEST_SET_SIZE'
```
[the remark this answers is not in the document]
Comment (jakehlee) [reply models/spectf/report_spectf.py:75]: training_example.py

File: models/spectf/report_spectf.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    type=click.Path(exists=True, dir_okay=False),
+    help="Path to the YAML config for the dataloader.",
+    envvar=f'{ENV_VAR_PREFIX}_DATA_CONFIG'
+)
+@click.option(
+    "--model-config",
+    required=True,
+    type=click.Path(exists=True, dir_okay=False),
+    help="Path to the YAML config for the model.",
+    envvar=f'{ENV_VAR_PREFIX}_MODEL_CONFIG'
```
[the remark this answers is not in the document]
Comment (jakehlee) [reply models/spectf/report_spectf.py:60]: training_example.py

File: models/README.md [hunk 10]
```diff
@@ -32,4 +32,10 @@ $ python3 models/training_example.py --help
 │    --seed                     INTEGER    Seed.                                                                                                                                   │
 │    --help                                Show this message and exit.                                                                                                             │
 ╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
-```
\ No newline at end of file
+```
+
+## SpecTf
+
+`training_spectf.py` is a SpecTf implementation of the training script, and `report_spectf.py` is an example intended to demonstrate inference.
```
Comment (michaelkiper) [inline models/README.md]: Please put the full relative path from the repo root

**Comment by michaelkiper** [comment]
@jakehlee could you make a subdir in `models` that is `models/spectf` where the `models/report_spectf.py`, `models/training_example.py`, and `models/training_spectf.py` live?

Besides that I think that everything is fine to merge

**Comment by jakehlee** [comment]
> @jakehlee could you make a subdir in `models` that is `models/spectf` where the `models/report_spectf.py`, `models/training_example.py`, and `models/training_spectf.py` live?

I moved `models/report_spectf.py` and `models/training_spectf.py`, but not `models/training_example.py` since... it's not using spectf.
